signer/src/bitcoin/validation.rs

```diff
@@ -91,10 +104,121 @@ pub fn is_unique(package: &[TxRequestIds]) -> bool {
+
+        self.validate_max_mintable(ctx, &cache).await?;
+
+        Ok(cache)
+    }
+
+    async fn fetch_all_reports<C>(
+        &self,
+        ctx: &C,
+        btc_ctx: &BitcoinTxContext,
+    ) -> Result<ValidationCache, Error>
+    where
+        C: Context + Send + Sync,
+    {
+        let db = ctx.get_storage();
+        let mut cache = ValidationCache {
+            sbtc_limits: ctx.get_emily_client().get_limits().await?,
+            ..Default::default()
+        };
+
+        // Fetch all deposit reports and votes
+        for requests in &self.request_package {
+            for outpoint in &requests.deposits {
+                let txid = outpoint.txid.into();
+                let output_index = outpoint.vout;
+
+                let report_future = db.get_deposit_request_report(
+                    &btc_ctx.chain_tip,
+                    &txid,
+                    output_index,
+                    &btc_ctx.signer_public_key,
+                );
+                let Some(report) = report_future.await? else {
+                    return Err(InputValidationResult::Unknown.into_error(btc_ctx));
+                };
+
+                let votes = db
+                    .get_deposit_request_signer_votes(&txid, output_index, &btc_ctx.aggregate_key)
+                    .await?;
+
+                cache
+                    .deposit_reports
+                    .insert((&outpoint.txid, output_index), (report, votes));
+            }
+
+            // Fetch all withdrawal reports and votes
+            for id in &requests.withdrawals {
+                let report = db.get_withdrawal_request_report(
+                    &btc_ctx.chain_tip,
+                    id,
+                    &btc_ctx.signer_public_key,
+                );
+                let Some(report) = report.await? else {
+                    return Err(WithdrawalValidationResult::Unknown.into_error(btc_ctx));
+                };
+
+                let votes = db
+                    .get_withdrawal_request_signer_votes(id, &btc_ctx.aggregate_key)
+                    .await?;
+
+                cache.withdrawal_reports.insert(id, (report, votes));
+            }
+        }
+        Ok(cache)
+    }
+
+    async fn validate_max_mintable<'a, C>(
+        &self,
+        ctx: &C,
+        cache: &ValidationCache<'a>,
+    ) -> Result<(), Error>
+    where
+        C: Context + Send + Sync,
+    {
+        let sbtc_supply = ctx
+            .get_stacks_client()
+            .get_sbtc_total_supply(&ctx.config().signer.deployer)
+            .await?;
+        let total_cap = cache.sbtc_limits.total_cap().unwrap_or(Amount::MAX_MONEY);
+        let max_mintable = total_cap
+            .checked_sub(sbtc_supply)
+            .unwrap_or(Amount::ZERO)
+            .to_sat();
+        cache
+            .deposit_reports
+            .values()
+            .try_fold(0u64, |acc, (report, _)| {
+                acc.checked_add(report.amount)
```

Comment:
Do we need to account for max fee here?

Note that there's other places where we don't account for max_fee, but I'll wait on commenting on those until we flesh out if it's needed or not in the previous comments.